### ISSUE TITLE
Typed widgets?

### ISSUE THREAD

**Issue by rafaqz** [piever/InteractBase.jl#27]
In Interact `slider()` produces an `Interact.Slider`.
InteractNext just outputs WebIO.Scope for every widget. If this is replacing InteractNext, what is the plan here?

Not having separate types means post-processing of widgets is difficult. I have a package that makes widgets and an interface from anything you throw at it, and reapplies signals back to the data, to save me rebuilding the interface every time I change my models:
https://github.com/rafaqz/AutoInteract.jl

But to make the layout passable you have to separate checkboxes/toggles from sliders etc somehow. In Interact that's easy, in InteractNext (and maybe this package?) it doesn't seem possible.

**Comment by piever** [comment]
At the moment widgets return either a `WebIO.Node` or a `WebIO.Scope`. However, I do think there's value in having a specialized `Widget` type (it could store the underlying `Scope` and `Node` and the primary observable, while rendering like the underlying Node. However, I think having a separate type per widget is a bit too much. I'd propose:

```julia
struct Widget{T}
    type::T
   node::Union{WebIO.Node, WebIO.Scope}
   scope::WebIO.Scope
   primary_obs::Observable
end
```

and you would dispatch on `f(x::Widget{Val{:slider}})`

Btw, could you share the code you use for automatic layout? There are quite a few options to create layout from InteractBase widgets and I was curious what would be perceived as most practical/intuitive (and also, how easy it'd be to recreate what you're doing in Interact in InteractBase).

**Comment by rafaqz** [comment]
For my case that doens't really add anything usefull. Whats really needed for post build customisation is a way of dispatching on widget type and there is no differentiation between widget types with just Widget. 

They could have the same fields and inherit from AbstractWidget so you don't need any extra methods.

**Comment by rafaqz** [comment]
Unless you mean type T is a subtyping? Sorry I missed that. T could just be a :symbol

**Comment by rafaqz** [comment]
It doesn't do great layout, if that's what your looking for! 
It can be tweaked with type-specific methods for vbox/hbox, a spreading algorythm, grids for matrices etc... and judicious CSS. But mostly it's just a free layout so I don't have to update layouts every time I change models around, so I can just play with the changes and plot whatever I want without writing code.

**Comment by rafaqz** [comment]
It also already works in InteractNext, if that wasn't clear... It just doesn't do layout so well because of WebIO.Scope...

To simplify things, I could cut out all the plotting parts and have AutoInteract and AutoPlot. 

I'm thinking maybe would also be a bad idea to tie it to InteractBase specifically, as all the recursive archicture doesn't care about the endpoint, I just use `@requires` to choose Interact or InteractNext, and that could be anything.

**Comment by piever** [comment]
Actually there is some converence: Interact will be ported to be a metapackage including InteractBase and the various CSS frameworks (InteractBulma, etc...), see [#226](https://github.com/JuliaGizmos/Interact.jl/pull/226), whereas InteractNext will become the "material design backend" for InteractBase, see [#18](https://github.com/JuliaGizmos/InteractNext.jl/pull/18), so in the long term you only need to support InteractBase and everything else will follow.

I'm working on the typing in [#28](https://github.com/piever/InteractBase.jl/pull/28), it actually does not require such a massive change and I think overall improves the package design so it may be possible to merge soonish.

Concerning the project per se, note that both `@manipulate` (as well as the `widget` function, to create widgets automatically from a bunch of types - `Range`, `Bool`, `String`, `Array`, `Color`, `Date`...) and (in the case of tabular data) [Sputnik](https://github.com/piever/Sputnik.jl) exist for creating GUIs automatically, so maybe it's worth to see how we can coordinate.

**Comment by rafaqz** [comment]
I was wondering if merging those efforts was the plan.

Sputnik looks good! I think there could be some productive exchange. I should point out that there are some differences in what we are calling automatic and what our data sources are. AutoInteract was made for making controls out of complex heterogeneous model parameters that might be in hierarchical composed types, tuples, arrays, matrices, and contain other unrelated cruft. It knows absolutely nothing about their structure. It's fundamentally messy and situation specific, part of the reason its kind of a widget toolkit more than a finished GUI, but you're encouraging me to make it something that just works, when passed a suitable solver/plotter method. 

Sputnik seems aimed at data science, which is a lot cleaner and more systematic, and you might not need to do any tree walking looking for sliders! So I'm not sure how much actual crossover there will be except the actual display tools, which I could definitely learn a lot about from Sputnik...

**Comment by rafaqz** [comment]
And thanks for getting onto typed widgets! Looks good. I think its definitely better to have a little more structure than raw WebIO.Scope.

**Comment by piever** [comment]
Fixed by #28

**Comment by piever** [comment]
I'm actually not going to keep the information of the widget in the type parameter as that can be dangerous in terms of compilation times: if you need to dispatch (but using conditionals on `widgettype(ui)` should be fine for most purposes), you can always do:

```julia
f(Val{widgettype(ui)}(), ui, ...)
```
